polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp

```diff
@@ -134,13 +133,79 @@ void UniformityAnalysis::visitOperation(
   // A memory side effects free operation that has uniform operands yields
   // uniform result(s).
   if (isMemoryEffectFree(op)) {
-    LLVM_DEBUG(llvm::dbgs() << "Operation is memory effect free\n");
+    LLVM_DEBUG(llvm::dbgs() << "UA: Operation is memory effect free\n");
     return propagateAllIfChanged(results, Uniformity::getUniform());
   }
 
   return analyzeMemoryEffects(op, operands, results);
 }
 
+void UniformityAnalysis::visitNonControlFlowArguments(
+    Operation *op, const RegionSuccessor &successor,
+    ArrayRef<UniformityLattice *> argLattices, unsigned firstIndex) {
+  LLVM_DEBUG(llvm::dbgs() << "UA: Visiting non-ctrl-flow args of operation: "
+                          << *op << "\n");
+
+  // If the lattice on any operand isn't yet initialized, bail out.
+  if (anyOfUniformityIsUninitialized(op->getOperands())) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "UA: Operand(s) uniformity not yet initialized\n");
+    return;
+  }
+
+  // Infer the uniformity of the loop IV by analyzing the loop bounds and step.
+  if (auto loop = dyn_cast<LoopLikeOpInterface>(op)) {
```

Comment:
what do you think of creating a visit function for loop, and returns optional Uniformity.
if return value is std::nullopt, then `return SparseDataFlowAnalysis::visitNonControlFlowArguments(op, successor, argLattices, firstIndex);`, else `return propagateAllIfChanged(*iv, *result);`?

Reply:
done